packages/devtools_app/lib/src/trees.dart

```diff
@@ -90,6 +90,13 @@ class TreeNode<T extends TreeNode<T>> {
     _isExpanded = true;
   }
 
+  /// Whether this node should be shown in the tree.
+  ///
+  /// When using this, consider caching the value. It is O([level]) to compute.
+  bool shouldShow() {
+    return parent == null || (parent.isExpanded && parent.shouldShow());
```

Comment:
this is O(depth-tree).
We need to cache this/bring more functionality from the inspector subclass of TreeNode into the root class as we now have more common concerns in all implementations.

Reply:
I believe that for any node, its level is the number of checks it needs to make to determine if it should show.

Bringing up functionality from the inspector's tree nodes sounds like a useful refactoring to me.